### PR TITLE
Implement composite metric handler and new metrics

### LINKED ISSUE #31: Moleculegen 1.1 Update
## Motivation
In this topic, we seek to introduce new and previously unobserved by the package ideas, which will likely become a part of moleculegen's enhancing functionality.

The issue will be updating by proposing ideas, and if the enhancement is feasible, the corresponding pull requests will mention this issue.

### Callbacks
- [x] The *Generator* callback to generate and optionally evaluate compounds every chosen epoch.
- [x] A phys.chem descriptor projection w/ dimensionality reduction. Visualizes training vs validation data distribution.

### Data
- [x] Include `augment` option in `moleculegen.data.SMILESDataset`.
- [x] Implement the *SMILESTargetDataset* compatible w/ both Stage 1 and Stage 2 procedures.
- [x] Pre-filtering of `moleculegen.data.SMILESVocabulary` attributes, including `corpus`.
- [x] An option to treat the subsequences enclosed in square brackets as separate tokens.
- [x] Implement a random sampling.
- [x] Accept *offset* parameter, which randomly selects starting position of SMILES string during sampling.

### Description
- [x] Create separate utilities to check and convert SMILES strings into molecules.
- [x] Refactor `moleculegen.description.get_descriptors*` correspondingly.
- [x] Scikit-learn-compatible transformers should accept RDKit molecules, not SMILES strings, in order to prevent multiple conversions on the same data during scikit-learn pipeline calculations.
- [x] Scikit-learn-compatible RDKit Descriptor transformer.
- [x] Scikit-learn-compatible *Internal Tanimoto distance* matrix transformer.

### Estimation
- [x] Add an option for fine-tuners to freeze embedding and encoder blocks during training.
- [x] Add an option for the main model to share embedding parameters w/ a decoder.
- [x] `moleculegen.estimation.SMILESRNN` has separate input-to-hidden and hidden-to-hidden initializers.
- [x] Implement SMILESGPT b75b59d
- [X] Implement SMILESCNN c7de81c

### Generation
- [x] Refactor `moleculegen.generation.GreedySearch`: a new generator *BaseSearch* w/ more options (hidden state initialization, `normalizer` does not require named parameters) and mxnet-only API.
- [x] The *Argmax (greedy) Search* generator (inherits *BaseSearch*).
- [x] The *Softmax Search* generator (inherits *BaseSearch*).
- [x] *Top-p* (nucleus) sampling.
- [x] *Top-k* sampling.

### Metrics
- [x] A composite metric to evaluate different metrics on the same training and validation data.
- [x] The *Novelty* metric to calculate the rate of novel compounds.
(`moleculegen.evaluation.RAC(count_unique=False)` calculates novel *and* valid compounds)
- [x] The *Uniqueness* metric to calculate the rate of unique compounds among generated ones.
- [x] The *Validity* metric to calculate the rate of valid compounds.
- [x] The *KL Divergence* of physchem descriptors.
- [x] *Internal Diversity* of generated compounds using Tanimoto similarity matrix.
- [x] *Nearest Neighbor Tanimoto Similarity* of generated compounds vs validation set.

### Miscellaneous
- [x] Logger to save callback messages.

### PR DESCRIPTION
**Pull request by sanjaradylov**
Resolves #31

- [x] `moleculegen.evaluation.CompositeMetric` - mutable sequence of metrics;
- [x] `moleculegen.evaluation.Novelty` - rate of novel compounds;
- [x] `moleculegen.evaluation.Uniqueness` - rate of unique compounds;
- [x] `moleculegen.evaluation.Validity` - rate of valid compounds.

*Notes*: `moleculegen.evaluation.RAC` was used to calculate all novelty, uniqueness, and validity simultaneously, and aggregate final result. Now we can monitor them separately (recommended).

### CODE REVIEW

File: moleculegen/evaluation/metric.py [hunk 1]
```diff
@@ -88,6 +110,303 @@ def reset(self):
...
+
+    def add(self, metric, *metrics):
+        """Append metric(s) to the sequence.
+
+        Parameters
+        ----------
+        metric : Metric
+        metrics : tuple of Metric, default ()
+        """
+        self._metrics.extend(itertools.chain((metric,), metrics))
```
Comment (sanjaradylov) [inline moleculegen/evaluation/metric.py]: For every metric in the chain, check if it is of type `moleculegen.evaluation.Metric`.

File: moleculegen/evaluation/metric.py [hunk 2]
```diff
@@ -88,6 +110,303 @@ def reset(self):
...
+        ------
+        TypeError
+            If `index` is out of range.
+            If `metric` is not of type `Metric`.
+        """
+        if not isinstance(metric, Metric):
+            raise TypeError(f'value must be of type Metric, not {type(metric)}')
+
+        try:
+            self._metrics.__setitem__(index, metric)
```
Comment (sanjaradylov) [inline moleculegen/evaluation/metric.py]: Use the operator instead of the dunder method.

**Review by sanjaradylov — COMMENTED** [review]
See comments.

**Review by sanjaradylov — COMMENTED** [review]
LGTM
